geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/LuceneQueryWithDifferentVersions.java

```diff
@@ -0,0 +1,101 @@
+
+public class LuceneQueryWithDifferentVersions
+    extends LuceneSearchWithRollingUpgradeDUnit {
+
+  // 2 locator, 2 servers
+  @Test
+  public void luceneQueryCannotBeExecuted()
+      throws Exception {
+    final Host host = Host.getHost(0);
+    VM locator1 = host.getVM(oldVersion, 0);
+    VM locator2 = host.getVM(oldVersion, 1);
+    VM server1 = host.getVM(oldVersion, 2);
+    VM server2 = host.getVM(oldVersion, 3);
+
+    final String regionName = "aRegion";
+    RegionShortcut shortcut = RegionShortcut.PARTITION_REDUNDANT;
+    String regionType = "partitionedRedundant";
+
+    int[] locatorPorts = AvailablePortHelper.getRandomAvailableTCPPorts(2);
+    locator1.invoke(() -> DistributedTestUtils.deleteLocatorStateFile(locatorPorts));
+    locator2.invoke(() -> DistributedTestUtils.deleteLocatorStateFile(locatorPorts));
+
+    String hostName = NetworkUtils.getServerHostName(host);
+    String locatorString = getLocatorString(locatorPorts);
+    try {
+      locator1.invoke(
+          invokeStartLocator(hostName, locatorPorts[0], getLocatorPropertiesPre91(locatorString)));
+      locator2.invoke(
+          invokeStartLocator(hostName, locatorPorts[1], getLocatorPropertiesPre91(locatorString)));
+      invokeRunnableInVMs(invokeCreateCache(getSystemProperties(locatorPorts)), server1, server2);
+
+      // Locators before 1.4 handled configuration asynchronously.
+      // We must wait for configuration configuration to be ready, or confirm that it is disabled.
+      locator1.invoke(
+          () -> await().untilAsserted(() -> await().untilAsserted(() -> {
+            assertThat(InternalLocator.getLocator().isSharedConfigurationRunning()).isTrue();
+          })));
+      locator2.invoke(
+          () -> await()
+              .untilAsserted(() -> await().untilAsserted(() -> {
+                assertThat(InternalLocator.getLocator().isSharedConfigurationRunning()).isTrue();
+              })));
+
+      server1.invoke(() -> createLuceneIndex(cache, regionName, INDEX_NAME));
+      server2.invoke(() -> createLuceneIndex(cache, regionName, INDEX_NAME));
+
+      invokeRunnableInVMs(invokeCreateRegion(regionName, shortcut.name()), server1, server2);
+
+      putSerializableObjectAndVerifyLuceneQueryResult(server1, regionName, 10, 0,
+          10, server1, server2);
+      locator1 = rollLocatorToCurrent(locator1, hostName, locatorPorts[0], getTestMethodName(),
+          locatorString);
+
+      locator2 = rollLocatorToCurrent(locator2, hostName, locatorPorts[1], getTestMethodName(),
+          locatorString);
+
+      server1 = rollServerToCurrentCreateLuceneIndexAndCreateRegion(server1, regionType, null,
+          shortcut.name(), regionName, locatorPorts, reindex);
+
+      Throwable thrown = catchThrowable(() -> {
+        putSerializableObjectAndVerifyLuceneQueryResult(server2, regionName, 20, 15,
+            25, server2);
+      });
+
+      assertThat(thrown.getCause()).isInstanceOf(AssertionError.class);
```

Comment:
This is the only assertion that this test makes, and it's very vague. Would it be possible to be more specific in terms of what exactly the test expects to fail, and where? Instead of asserting that some other assertion from `putSerializableObjectAndVerifyLuceneQueryResult()` fails, it would be better to have a specific assertion for this test showing exactly where and how we expect the query verification to fail.

Reply:
Thi test checks that it is impossible to do lucene query with different version of the servers. It will failed in ` void verifyLuceneQueryResults(String regionName, int expectedRegionSize)` while doing the following assertion:
`    assertTrue((Boolean) luceneService.getClass()
        .getMethod("waitUntilFlushed", String.class, String.class, long.class, TimeUnit.class)
        .invoke(luceneService, INDEX_NAME, regionName, 60, TimeUnit.SECONDS));`
I'm not sure how to change this assertion in the test to verify this assertion will failed.

Reply:
It looks like the failure is happening before the Lucene query even begins, since `waitUntilFlushed()` is returning false rather than the expected true. This appears to be because the wait times out, and I also see logging related to stuck Pooled Waiting Message Processor threads, and the message "During normal processing, unsuccessfully dispatched 1 events (batch #0)" being logged repeatedly. This seems like unwanted behaviour, that should be looked into further, as we shouldn't be getting stuck like this if the safeguards in this PR are working as intended.

Reply:
Yes, I agree that this message "During normal processing,..." should be avoided. I will try to investigate how to fix this part.